gbpservice/nfp/configurator/agents/vpn.py

```diff
@@ -0,0 +1,275 @@
+               'method': 'ipsec_site_conn_deleted',
+               'data': {'context': context,
+                        'resource_id': resource_id}
+               }
+        self.notify._notification(msg)
+
+
+class VPNaasRpcManager(agent_base.AgentBaseRPCManager):
+    """
+    APIs for receiving RPC messages from vpn plugin.
+    """
+    RPC_API_VERSION = '1.0'
+    target = messaging.Target(version=RPC_API_VERSION)
+
+    def __init__(self, conf, sc):
+        """Instantiates child and parent class objects.
+
+        Passes the instances of core service controller and oslo configuration
+        to parent instance in order to provide event enqueue facility for batch
+        processing event.
+
+        :param sc: Service Controller object that is used for interfacing
+        with core service controller.
+        :param conf: Configuration object that is used for configuration
+        parameter access.
+
+        """
+
+        super(VPNaasRpcManager, self).__init__(conf, sc)
+
+    def vpnservice_updated(self, context, kwargs):
+        arg_dict = {'context': context,
+                    'kwargs': kwargs}
+        ev = self.sc.new_event(id='VPNSERVICE_UPDATED', data=arg_dict)
+        self.sc.post_event(ev)
+
+
+class VPNaasEventHandler(object):
+    """
+    Handler class for demultiplexing vpn configuration
+    requests from VPNaas Plugin and sending to appropriate driver.
+    """
+    def __init__(self, sc, drivers):
+        self.sc = sc
+        self.drivers = drivers
+        self.needs_sync = True
+        self.context = ctxt.get_admin_context_without_session()
+        self.plugin_rpc = VpnaasRpcSender(
+            self.context,
+            self.sc)
+
+    def _get_driver(self):
+
+        driver_id = const.SERVICE_TYPE
+        return self.drivers[driver_id]
+
+    def handle_event(self, ev):
+        try:
+            msg = ("Worker process with ID: %s starting "
+                   "to handle task: %s of topic: %s. "
+                   % (os.getpid(), ev.id, const.VPN_GENERIC_CONFIG_RPC_TOPIC))
+            LOG.debug(msg)
+
+            driver = self._get_driver()
+            self.vpnservice_updated(ev, driver)
+        except Exception as err:
+            LOG.error("Failed to perform the operation: %s. %s"
+                      % (ev.id, str(err).capitalize()))
+        finally:
+            self.sc.event_done(ev)
+
+    def vpnservice_updated(self, ev, driver):
+        context = ev.data.get('context')
+        kwargs = ev.data.get('kwargs')
+        LOG.debug(_("Vpn service updated from server side"))
+
+        try:
+            # in future if the vpnservice service function changes
+            # then one should get the below function with getattr function
+            driver.vpnservice_updated(context, kwargs)
+        except Exception as err:
+            LOG.error("Failed to update VPN service. %s"
+                      % str(err).capitalize())
+
+        reason = kwargs.get('reason')
```

Comment:
If VPN driver fails to delete, we should not send ipsec_site_conn_deleted notification right?